feat: Pre-release patches toggle

kitadai31 opened the pull request:
Add a toggle for pre-release patches to the alternative source settings.
Closes: #821

<img src="https://github.com/ReVanced/revanced-manager/assets/90122968/2de2f839-031c-427d-a403-a32e7b0b80af" width="240">

<img src="https://github.com/ReVanced/revanced-manager/assets/90122968/8b55d5a8-921e-4704-8fec-eb086384ffc0" width="240">

<img src="https://github.com/ReVanced/revanced-manager/assets/90122968/caab633d-a514-47e4-b22a-48d4a68d057b" width="240">

This PR also fixes an problem where pre-release Integrations are being used when using alternate sources.
At this time, Patches are referring `GitHubAPI#getLatestPatchesRelease()` that doesn't include pre-releases, but Integrations are referring `GitHubAPI#getLatestRelease()` that includes pre-releases.
This problem actually caused patched apps to crash recently in crimera/piko patches (an alternative source) when they released a pre-releases.
(Actually, this is the main motivation)

`getLatestReleaseFile()` in `github_api.dart` was removed because it's unused

Ushie commented:
Why limit it to alternative sources?

You can pass ?dev=true to achieve the same with the ReVanced API

kitadai31 commented:
I already tried ?dev=true at ReVanced API, but it didn't effect.
https://api.revanced.app/v2/revanced-patches/releases/latest?dev=true
So, I thought it wasn't implemented yet at first.
That's why it is limited to GitHub API only.  I didn't know the cache issue.
If the cache issue will be fixed, I think I should wait for it.

> Are changelogs/updates considered as well?

Yes. These lines change the patches update dialog according to the settings.
https://github.com/ReVanced/revanced-manager/blob/e88e997436d8692023d49480089fe7d71538492b/lib/ui/views/home/home_viewmodel.dart#L482-L490

brian6932 commented:
One thing I noticed after building this branch (rebased) was that RevancedManager, gets stuck on `Loading...`, patches work fine though. Is that just a quirk of debug builds, or an issue with the PR?

nivranaitsirhc commented:
> One thing I noticed after building this branch (rebased) was that checking for `Revanced Manager` updates gets stuck on `Loading...`, patches work fine though. Is that just a quirk of debug builds, or an issue with the PR?

Can you share the apk here please

brian6932 commented:
@nivranaitsirhc sure https://i.alienpls.org/ReVanced.apk

nivranaitsirhc commented:
> @nivranaitsirhc sure https://i.alienpls.org/ReVanced.apk

It works like a charm! Thanks!

![Screenshot_20240420-191932_ReVanced Manager Debug.png](https://github.com/ReVanced/revanced-manager/assets/13590148/cb4d926e-ead2-4ce2-b5c0-c8d51b2014a8)

![Screenshot_20240420-191959_ReVanced Manager Debug.png](https://github.com/ReVanced/revanced-manager/assets/13590148/e0707194-b685-44cd-b05c-37aa875a4047)

![Screenshot_20240420-193744_ReVanced Manager Debug.png](https://github.com/ReVanced/revanced-manager/assets/13590148/97d47207-0a64-4fc5-9b50-fac939f65b95)

![Screenshot_20240420-194113_YouTube.png](https://github.com/ReVanced/revanced-manager/assets/13590148/b0e76957-46d6-4a24-ab2b-b94427eaabd9)

kitadai31 commented:
Waiting for the parameter issue in ReVanced API to be fixed

I have some questions

1. I added the ReVanced API v2 related logic in revanced_api.dart, but should I split the file into two files like `revanced_api_v2.dart` ?
2. I added a word "(dev)" in settings description because ReVanced calls pre-release versions “dev”. (40b1589)
Is this no problem?

Ushie commented on lib/services/github_api.dart:
```diff
@@ -37,6 +37,34 @@ class GithubAPI {
     }
   }
 
+  Future<Map<String, dynamic>?> getLatestReleaseWithPreReleases(
+    String repoName,
+  ) async {
+    try {
+      final response = await _dio.get('/repos/$repoName/releases?per_page=10');
+      final List<dynamic> releases = response.data;
+
+      /*
+      * Loop through all releases (including pre-releases) and return the latest
+      */
+      Map<String, dynamic>? latestRelease;
+      DateTime latestReleaseDate = DateTime.fromMillisecondsSinceEpoch(0);
+      for (final release in releases) {
+        final DateTime releaseDate = DateTime.parse(release['published_at']);
+        if (releaseDate.isAfter(latestReleaseDate)) {
+          latestReleaseDate = releaseDate;
+          latestRelease = release;
+        }
+      }
+      return latestRelease;
+    } on Exception catch (e) {
+      if (kDebugMode) {
+        print(e);
+      }
+      return null;
+    }
+  }
+
```
Maybe instead of making a new function, add a parameter to the existing one instead

Ushie reviewed (changes requested):
@oSumAtrIX Does v3 suffer from the same issues regarding the ?dev parameter, supporting the API is a prerequisite for this PR

oSumAtrIX commented:
A ?dev query parameter does not exist and is not planned to be added. The TLDR is that the dev branch is not a controlled release channel like a "beta" channel would be.

Ushie commented:
The dev branch has nothing to do with this, the flag can be renamed to ?prereleases instead of ?dev because it simply enables returning prereleases

oSumAtrIX commented:
The name does not change that pre releases are not a controlled production release channel. Any further discussion should happen in the API repo and not in this PR

kitadai31 commented:
Yeah, I've heard that ReVanced API v2 was deprecated.
And v3 doesn't have toggles for pre-releases.

So, I think it's enough to revert this PR to the beginning. (Pre-release requires alternative sources to be enabled, https://github.com/ReVanced/revanced-manager/pull/1785#issue-2205094068)
